text/0000-process-specific-env.md

```diff
@@ -0,0 +1,81 @@
+
+```
+# defined processes
+/cnb/lifecycle/launcher debug # NODE_OPTIONS=--inspect=0.0.0.0:9229
+/cnb/lifecycle/launcher web # no NODE_OPTIONS defined
+
+# ad-hoc
+/cnb/lifecycle/launcher -- npm install # no NODE_OPTIONS defined
+/cnb/lifecycle/launcher -- bash # no NODE_OPTIONS defined
+```
+
+`<layer>/profile.d/debug/debug.sh` would work similarly and be sourced by Bash before launching the `debug` command, but not when executing any other process.
+
+# What it is
+[what-it-is]: #what-it-is
+
+Each env or profile layer dir can have subdirs like `<layer>/env.launch/web/` and `<layer>/profile.d/web/`. These behave the same way as `<layer>/env` but are specific to the launch of the `web` process.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+Lifecycle would use the identifier of the intended launch process to bring in additional `env` and `profile.d` files. The launcher would sort these process specific values _after_ less specific values to allow [Environment Variable Modification Rules](https://github.com/buildpacks/spec/blob/master/buildpack.md#environment-variable-modification-rules) to apply as expected.
+
+The environment modification rules should be updated to reflect that sort order for `.prepend`, `.append`, and `.override` to reflect that process specific values are more processed in the most overriding order. The following shows an example of processing order for a given layer.
+
+```
+<layers>/<layer>/env/NODE_OPTIONS.append
+<layers>/<layer>/env.launch/NODE_OPTIONS.append
+<layers>/<layer>/env.launch/web/NODE_OPTIONS.append
+```
+
+Example spec update:
+
+- Environment variable file contents originating in the same layer MUST be sorted such that file contents in <layers>/<layer>/env.launch/<process>/ precede file contents in <layers>/<layer>/env.launch/, which must precede <layers>/<layer>/env/.
+- Environment variable file contents originating in the same layer MUST be sorted such that file contents in <layers>/<layer>/env/ precede file contents in <layers>/<layer>/env.build/ or <layers>/<layer>/env.launch/ which must preced file contents in <layers>/<layer>/env.launch/<process>/.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Additional complexity would be introduced into the launcher.
+- Introducing process specific directory structure for `env` and `profile.d` but not `bin` and `lib`, for example, could be confusing for buildpack authors.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- Do Nothing. Buildpack authors can continue work around some of the restrictions today by embedding process specific `env` values into the launch processes instead of using the `env` directory.
+
+# Prior Art
+[prior-art]: #prior-art
+
+- Copied from [here](https://github.com/buildpacks/rfcs/pull/72) and updated to include additional details and examples.
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+- Should this idea be extended to `bin`, `lib`, and any other buildpack output intended for launch?

```

Comment:
I'm not sure what the use case is? @jabrown85

Reply:
A user might want to add a credential helper binary for `web` but not have that accessible for `debug`. They could put the binary at `bin/web/credential-helper` to achieve this if we decided to allow this.

Reply:
Is that a different image at that point?

Reply:
I'm not sure. I think it's reasonable to think that perhaps a `web` process is restricted to having access to something in `bin/lib` that perhaps you wouldn't want to be immediately accessible for non-defined launcher executions `/cnb/lifecycle/launcher -- bash` for instance.

Reply:
Yeah, I guess without changing the image it's still technically there. I'm happy to mark this as out of scope of the RFC unless you feel strongly about it.